### DISCUSSION TITLE
Queries for Dynamic Components

### DISCUSSION THREAD

**Discussion by zicklag** [bevyengine/bevy#635] [Q&A] [answered]
Hey everybody! I'm currently working on laying the foundation for dynamic components in systems in Bevy. I've already started work on that in #623. Now I would like to get some ideas, though, particularly around how to refactor Queries in a way that will support making a query based on a runtime-determined component ID.

Let's get some context.

## Component ID's

The first step that I did in #623 was I converted all of the uses of `TypeId` to identify components in the world storage and switched to to a `ComponentId` enum that can represent either `RustTypeId(TypeId)` or `ExternalId(u64)`. That commit leaves us storing things based on `ComponentId`, but the rest of the ECS still pretty much uses `TypeId` and just calls `TypeId::of::<T>().into()` to convert to component IDs everywhere that it's required. So the ECS functionality as it stands right now is pretty much identical to how it was before, but now we are internally able to represent the storage of custom components.

## The Goal

So the current goal is to be able to define a system without knowing which components it's going to access at compile time. This system needs to be able to make a `Query` that can query based on runtime parameters instead of the usual compile type type parameters.

For instance, a query in Bevy looks like `Query<(&mut Pos, &Vel)>`. The Query trait is actually very clever and is defined as:

```rust
pub trait Query {
    type Fetch: for<'a> Fetch<'a>;
}
```

And the `Fetch` trait is:

```rust
pub trait Fetch<'a>: Sized {
    type Item;

    fn access(archetype: &Archetype) -> Option<Access>;

    fn borrow(archetype: &Archetype);
    unsafe fn get(archetype: &'a Archetype, offset: usize) -> Option<Self>;
    fn release(archetype: &Archetype);

    unsafe fn should_skip(&self) -> bool {
        false
    }
    unsafe fn next(&mut self) -> Self::Item;
}
```

So the gist of this is that the `Query` essentially exists, for the most part only as a type! The issue I have with this centers around the `access`, `borrow`, and `release` functions of the `Fetch` trait. Because there's no `&self` parameter to any of those functions, the only thing that anything implementing `Fetch` has to go off of to determine which archetypes that it should access is it's type.

This issue is that, when queries need to be instantiated at runtime, we need to be able to implement the `access` function, for instance, in terms of a concrete, initialized `ComponentId`, **not** based on a type parameter.

### Modify `Fetch` Trait to Allow Passing State Into `access`, `borrow`, and `release`

My first thought was to modify the `Fetch` to allow passing in a `ComponentId` somehow. That's essentially a requirement as far as I can think. We **need** to be able to modify the behaviour the `access`, `borrow`, and `release` based on a `ComponentId` ( or more likely a list list of `ComponentId`s or some other wrapping data structure encapsulating the `ComponentId` and the desired `Access` ).

I experimented a bit, but there seems to be no clean way to introduce the `ComponentId` information to the existing `Fetch` types. It just drastically modifies the way that they currently work and makes it rather messy.

### Implement a New `DynamicQuery`

The other idea is to just re-implement all of the `Query` logic in terms of a new `DynamicQuery` that accomplishes the same functionality but uses runtime-initialized structs to determine the query parameters. I was wanting to avoid that because a lot of the functionality would probably be duplicated.

If we implemented a `DynamicQuery` we could possibly re-implement `Query` in terms of the new `DynamicQuery`, to avoid code duplication, but I'm not sure if that would work out or not.

---

I'm not fully settled on a direction and I would like to get ideas or suggestions on what route to take. I'm leaning towards implementing `DynamicQuery` as I feel like mutilating the current type-based `Fetch` and `Query` traits is not going to end well.

I think I'm going to attempt implementing `DynamicQuery` and the implementing `Query` on top of `DynamicQuery` to share bulk of the logic other than the actual creation of the `DynamicQuery` from the type. So essentially you could construct a `Query` like you do today, but maybe it would implement `Into<DynamicQuery>` and then `DynamicQuery` is where all of the query logic, that is today implemented on `Query`, would go.

What do you think?

**Comment by cart** [comment]
This is hard. I think in the short term a new DynamicQuery specifically for this case is the right approach, even if it means a bit of code duplication.

Medium term, theres a chance we'll need to implement a "stateful query" anyway to support 'system execution order independent change tracking'. The design we've discussed so far would look something like this:

```rust
// state contains a "last change count" for Changed<T> 
world.stateful_query::<(Changed<T>, &U)>(state);
```

Not sure if theres a way to reconcile that with the requirements here / if we should.

Another thing to consider is that Fetch is a "hot spot" and any change has a good chance of regressing performance.

**Reply by zicklag** [reply]
> I think in the short term a new DynamicQuery specifically for this case is the right approach, even if it means a bit of code duplication.

OK, that sounds good. Even if it isn't final, it allow me to keep moving on the scripting experimentation and we can come back and refine it later.

> Medium term, theres a chance we'll need to implement a "stateful query" anyway to support 'system execution order independent change tracking'...Not sure if theres a way to reconcile that with the requirements here / if we should.

Yeah stateful queries does look like it would open a door for what we need there, which would be good. Probably another good reason for me to wait on trying to redesign that. 

> Another thing to consider is that Fetch is a "hot spot" and any change has a good chance of regressing performance.

Yeah, I figured. I was a bit afraid to touch it for that reason. It's pretty cool how the queries almost "don't exist" at runtime, but it makes it difficult for dynamism. :smile:

**Comment by zicklag** [comment]
OK, I just noticed something really weird. So I decided to mess around with `Fetch` and run the `ecs_bench` on it to make sure I didn't slow anything down, and for some reason it got *faster* when I added an extra `State` type parameter set to the unit type `()`.

These are the changes I applied:

https://github.com/katharostech/bevy/commit/efa4ec363a3b95914b7233848f76b2fa6cbd9f58 ( Sorry for the noise, but some of those changes are somewhat unrelated. The interesting stuff is all in the `query.rs` file. )

According to @cart's [`ecs_bench`](https://github.com/cart/ecs_bench) after my changes, which should have made it slower if anything, somehow makes it ~20% *faster*! It's only faster by ~15 nanoseconds, but its consistently faster on every comparison between the code before that commit and after that commit. On my machine the benchmarks seem to have about ±3% noise between identical compiles, so take that into account, but here is the benchmark **after** the changes:

```
Gnuplot not found, using plotters backend
pos_vel/bevy            time:   [1.0387 us 1.0599 us 1.0817 us]                          
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe
pos_vel/bevy_entity     time:   [922.38 ns 925.00 ns 928.16 ns]                                 
Found 17 outliers among 100 measurements (17.00%)
  1 (1.00%) low mild
  10 (10.00%) high mild
  6 (6.00%) high severe
pos_vel/bevy_foreach    time:   [925.34 ns 926.98 ns 928.72 ns]                                  
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) low mild
  5 (5.00%) high mild
  3 (3.00%) high severe
pos_vel/bevy_foreach_entity                                                                             
                        time:   [912.99 ns 914.99 ns 917.13 ns]
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) low mild
  8 (8.00%) high mild
  3 (3.00%) high severe
```

Switching back to the code **before** the commit and running the benchmarks:

```
pos_vel/bevy            time:   [1.1588 us 1.1612 us 1.1642 us]                          
                        change: [+22.843% +23.152% +23.488%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe
pos_vel/bevy_entity     time:   [1.2476 us 1.2561 us 1.2648 us]                                 
                        change: [+34.213% +35.018% +35.835%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild
pos_vel/bevy_foreach    time:   [942.89 ns 946.01 ns 949.63 ns]                                  
                        change: [-8.8439% -7.8616% -6.8525%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe
pos_vel/bevy_foreach_entity                                                                             
                        time:   [941.50 ns 949.15 ns 957.64 ns]
                        change: [-6.4771% -5.2201% -3.8825%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

```

The changes for bevy_foreach and bevy_foreach_entity seem to just be noise according to the runs I've done and how much they can vary without me actually changing the code. 

Does a change this small even make a difference? I don't know. I'm definitely no performance/benchmarking expert yet so I don't know if I'm missing anything, but I'm kind confused how it is faster when all I did is add another type parameter and a bunch of useless arguments.

---

Anyway, now the `Fetch` trait has optional state without slowing down the use-case where it doesn't need the state ( such as today's `Query` implementations ). I'm going to experiment a bit more with that to see if I can include the `ComponentId` in the state and create a `Fetch` implementation that I can build a `DynamicQuery` on.

**Reply by zicklag** [reply]
More info on the strange performance characteristics of that extra `State` type parameter to `Fetch`:

So in the commit I mention above that caused the performance increase, I somewhat sloppily constraint the new `State` type parameter to implementing `Default` and then passing in `&Default::default()` as the `state` for all of the functions that didn't have any state while setting `type State = ()`.

What would have been much cleaner, I thought, was to remove the `Default` constraint and just constraint `State=()` for all of our current `Fetch` implementations. ( see https://github.com/katharostech/bevy/commit/7791b6a1477d1a312e443aba31b3e41c8060dd55 ). For some reason though, this reset the benchmark results back to what they were *before I did **any** changes*.

So using the `Default` constraint for the `State` type parameter and passing in `&Default::default` for the state in all the functions is somehow faster than constraining `State=()` and just passing in `&()`. I'm confused, but for now I'll keep the faster one. :man_shrugging: 

---

I made some progress testing out making `QueryBorrowChecked` generic across `Fetch` implementations with different `State`s. Still didn't get anything fully working, but looks promising. I've never done much with lifetimes, but I'm learning stuff so this is interesting.

**Reply by zicklag** [reply]
I'll just keep posting my progress here:

I just managed to implement `QueryBorrowChecked` and with it 95% of the actual query logic, as generic over the the `State` type parameter of `Fetch` all while maintaining the ~20% iteration speed boost:

https://github.com/katharostech/bevy/commit/759d12a278e81197b868f6ad89a9e8765c8f285e

It makes the `QueryBorrowChecked` and related types rather ridiculously annotated, one of them with 3 lifetimes and two generic type parameters, but it works! Now I think I just have to implement `DynamicQuery` on top of a `DynamicFetch` implementation that filter components based on an array of a new `ComponentInfo` type.

**Reply by zicklag** [reply]
OK yeah, the `ecs_bench` results are changed by really odd things. If I simply comment out the benches for the foreach systems inside of the `bevy.rs` file the `bevy` and `bevy_entity` benchmarks are slowed down by 100%!

Don't know what the deal with that is.

**Reply by cart** [reply]
Yeah I've noticed similar behavior. It's sadly not enough to make systems run fast for one permutation. They need to run fast for "the average permutation".

**Reply by cart** [reply]
Some code is more "unstable" than other code.

**Comment by zicklag** [comment, marked as answer]
I was successfully able to introduce state into queries and establish a dynamic system! I've still got to do some cleanup work and I have to do more testing, but the existing bevy queries continue to work without changes and without a performance regression according to ECS bench run before and after my commits. Here is the current [branch](https://github.com/katharostech/bevy/tree/02-fetch-experiments). I'll be cleaning it up and moving it to #623.

Here's the example code that sets up a runtime-created system:

```rust
use std::time::Duration;

use bytemuck::{Pod, Zeroable};

use bevy::prelude::*;
use bevy_app::ScheduleRunnerPlugin;
use bevy_ecs::{
    Access, ComponentId, DynamicComponentAccess, DynamicComponentInfo, DynamicComponentQuery,
};

// Define our componens

#[derive(Debug, Clone, Copy)]
struct Pos {
    x: f32,
    y: f32,
}

#[derive(Debug, Clone, Copy)]
struct Vel {
    x: f32,
    y: f32,
}

// Implement `Pod` ( Plain 'ol Data ) and `Zeroable` for our components so that we can cast them
// safely from raw bytes later

unsafe impl Zeroable for Pos {}
unsafe impl Zeroable for Vel {}
unsafe impl Pod for Pos {}
unsafe impl Pod for Vel {}

/// Create a system for spawning the scene
fn spawn_scene(world: &mut World, _resources: &mut Resources) {
    #[rustfmt::skip]
    world.spawn_batch(vec![
        (
            Pos {
                x: 0.0,
                y: 0.0
            },
            Vel {
                x: 0.0,
                y: 1.0,
            }
        ),
        (
            Pos {
                x: 0.0,
                y: 0.0
            },
            Vel {
                x: 0.0,
                y: -1.0,
            }
        )
    ]);
}

/// Create a system for printing the status of our entities
fn info(mut query: Query<(&Pos, &Vel)>) {
    println!("---");
    for (pos, vel) in &mut query.iter() {
        println!("{:?}\t\t{:?}", pos, vel);
    }
}

fn main() {
    // A Dynamic component query which can be constructed at runtime to represent which components
    // we want a dynamic system to access.
    //
    // Notice that the sizes and IDs of the components can be specified at runtime and allow for
    // storage of any data as an array of bytes.
    let mut query = DynamicComponentQuery::default();

    // Set the first element of the query to be write access to our `Pos` component
    query[0] = Some(DynamicComponentAccess {
        info: DynamicComponentInfo {
            id: ComponentId::RustTypeId(std::any::TypeId::of::<Pos>()),
            size: std::mem::size_of::<Pos>(),
        },
        access: Access::Write,
    });

    // Set the second element of the query to be read access to our `Vel` component
    query[1] = Some(DynamicComponentAccess {
        info: DynamicComponentInfo {
            id: ComponentId::RustTypeId(std::any::TypeId::of::<Vel>()),
            size: std::mem::size_of::<Vel>(),
        },
        access: Access::Read,
    });

    // Create our dynamic system by specifying the name, the query we created above, and a closure
    // that operates on the query
    let pos_vel_system = DynamicSystem::new("pos_vel_system".into(), query, |query| {
        // Iterate over the query just like you would in a typical query
        for mut components in &mut query.iter() {
            // `components` will be an array with indexes corresponding to the indexes of our
            // DynamicComponentAccess information that we constructed for our query when creating
            // the system.
            //
            // Each item in the array is an optional mutable reference to a byte slice representing
            // the component data: Option<&mut [u8]>.

            // Here we take the mutable reference to the bytes of our position and velocity
            // components
            let pos_bytes = components[0].take().unwrap();
            let vel_bytes = components[1].take().unwrap();

            // Instead of interacting with the raw bytes of our components, we first cast them to
            // their Rust structs
            let mut pos: &mut Pos = &mut bytemuck::cast_slice_mut(pos_bytes)[0];
            let vel: &Vel = &bytemuck::cast_slice(vel_bytes)[0];

            // Now we can operate on our components
            pos.x += vel.x;
            pos.y += vel.y;
        }
    });

    App::build()
        .add_plugin(ScheduleRunnerPlugin::run_loop(Duration::from_secs(1)))
        .add_startup_system(spawn_scene.thread_local_system())
        .add_system(Box::new(pos_vel_system))
        .add_system(info.system())
        .run();
}

```

**Reply by zicklag** [reply]
OK, moved all the code to #623. Should probably do any further discussion there.
